src/LazyHydrate.js

```diff
@@ -0,0 +1,141 @@
+import React from 'react'
+import PropTypes from 'prop-types'
+
+import { StylesProvider, createGenerateClassName } from '@material-ui/core/styles'
+import { SheetsRegistry } from 'jss'
+
+const registries = []
+const generateClassName = createGenerateClassName()
+
+/*
+  This component renders the server side rendered stylesheets for the
+  lazy hydrated components. Once they become hydrated, these stylesheets
+  will be removed.
+*/
+export function LazyStyleElements() {
+  return (
+    <>
+      {registries.map((registry, index) => {
+        const id = `jss-lazy-${index}`
+        return <style key={id} id={id} dangerouslySetInnerHTML={{ __html: registry.toString() }} />
+      })}
+    </>
+  )
+}
+
+function LazyStylesProvider({ children }) {
+  const registry = new SheetsRegistry()
+  registries.push(registry)
+  return (
+    <StylesProvider
+      sheetsManager={new Map()}
+      serverGenerateClassName={generateClassName}
+      sheetsRegistry={registry}
+    >
+      {children}
+    </StylesProvider>
+  )
+}
+
+const isBrowser =
+  typeof window !== 'undefined' &&
+  typeof window.document !== 'undefined' &&
+  typeof window.document.createElement !== 'undefined'
+
+// Used for detecting when the wrapped component becomes visible
+const io =
+  isBrowser && IntersectionObserver
+    ? new IntersectionObserver(entries => {
+        entries.forEach(entry => {
+          if (entry.isIntersecting || entry.intersectionRatio > 0) {
+            entry.target.dispatchEvent(new CustomEvent('visible'))
+          }
+        })
+      })
+    : null
+
+function LazyHydrateInstance({ hydrated, ssrOnly, children, on, index, ...rest }) {
+  const childRef = React.useRef(null)
+  const [_hydrated, setHydrated] = React.useState(!isBrowser)
+
+  React.useEffect(() => {
+    if (ssrOnly || hydrated) return
```

Comment:
```suggestion
    if (ssrOnly || _hydrated) return
```

Reply:
I believe this is correct, we are checking if the passed `hydrated` prop is true, so they can control the component if desired.

Reply:
I will document that prop